### DISCUSSION TITLE
Installing multiphenicsx with a dockerfile

### DISCUSSION THREAD

**Discussion by bbbyyfff** [multiphenics/multiphenicsx#18] [Q&A] [answered]
Dear Community,

I am trying to use a docker file to create a personalised image that contains `multiphenicsx`. My docker file writes as follows:
```
FROM dolfinx/lab:nightly
LABEL description="nightly"

# Install dependencies for pyvista and related packages
RUN apt-get update && \
    apt-get install -y libgl1-mesa-dev xvfb && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*

# Install dependencies required for adding repositories
RUN apt-get update && \
    apt-get install -y software-properties-common && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*

# Pyvista ITKWidgets dependencies
RUN pip3 install --no-cache-dir itkwidgets ipywidgets matplotlib

# Install meshio
RUN pip3 install --no-cache-dir --no-binary=h5py h5py meshio 

# Additional python modules
RUN pip3 install --no-cache-dir sympy

# Install progress-bar
RUN pip3 install tqdm

# Clone multiphenicsx repository from github
# RUN git clone https://github.com/multiphenics/multiphenicsx.git /root/multiphenicsx

# Change directory to the cloned repository
# WORKDIR /root/multiphenicsx

# Install multiphenics

# RUN pip3 install scikit-build-core[pyproject]
# RUN pip3 install --check-build-dependencies --no-build-isolation .[tutorials]


EXPOSE 8888/tcp
ENV SHELL /bin/bash

ENTRYPOINT ["jupyter", "lab", "--ip", "0.0.0.0", "--no-browser", "--allow-root"]
```

If I don't install `multiphenicsx`, it works perfectly (I tried with a hyperelastic mixed problem), but when I include the lines that install `multiphenicsx`, it somehow removes the `ufl.VectorElement` in the `ufl` library.

I was wondering if I did something wrong and if there is a neat way of doing this. 

Thank you very much!

Kind regards,
Yifei

**Comment by bbbyyfff** [comment]
An update:

I have tried different combinitions of versions of dolfinx and multiphenicsx and dolfinx version = 0.7.0 and multiphenicsx version = 0.2.dev works. 

Using nightly image for dolfinx and git cloning directly from multiphenicsx github repository doesn't work for me as shown in the code in last post.

**Comment by francesco-ballarin** [comment, marked as answer]
Hi @bbbyyfff ,
* the removal of `ufl.VectorElement` was done upstream in https://github.com/FEniCS/ufl/pull/197. `multiphenicsx` tutorials were updated in https://github.com/multiphenics/multiphenicsx/commit/22479c60d2d2306ea06f8575c087f12925f17747: I guess you are running your own code, please make sure to update it similarly to that commit.
* you do not get an issue with `dolfinx/dolfinx:v0.7.0`, because the `ufl` change was done _after_ `v0.7.0` was released. However, be aware of https://multiphenics.github.io/installing.html#compatibility-with-upstream-releases when installing `multiphenicsx` on `dolfinx` releases.
* I don't expect this part
```
# Pyvista ITKWidgets dependencies
RUN pip3 install --no-cache-dir itkwidgets ipywidgets matplotlib
```
to be needed anymore. I moved the code related to interactive plotting to `viskex`, and that library automatically installs `pyvista` and its dependencies. Furthermore, `pyvista` has dropped support for `itkwidgets`, and a different plotting backend is now used.

Cheers,
Francesco

**Reply by bbbyyfff** [reply]
Hi Francesco!

Thank you very much! I should probably update the codes and images more often...


Yifei

**Reply by francesco-ballarin** [reply]
True, but _right now_ is not a good time :rofl: , see https://github.com/multiphenics/multiphenicsx/issues/19

**Reply by bbbyyfff** [reply]
I see your point now!
